### ISSUE TITLE
Double nested generic fails, Foo<Bar> works but Foo<List<Bar>>> fails

### ISSUE THREAD

**Issue by jntakpe** [micronaut-projects/micronaut-serialization#413]
### Expected Behavior

Given the following file: 
```kotlin
@Client("\${dummy.url}")
interface DummyClient {

    @Get("/api/dummy/wrapped/list")
    fun wrappedList(): ApiResponse<List<Dummy>>


    @Get("/api/dummy/wrapped/nested")
    fun wrappedNested(): ApiResponse<ApiResponse<Dummy>>

    @Get("/api/dummy/wrapped/simple")
    fun wrappedSimple(): ApiResponse<Dummy>

    @Get("/api/dummy/raw")
    fun simpleList(): List<Dummy>
}

@Serdeable
data class Dummy(val name: String)

@Serdeable
data class ApiResponse<T>(val content: T)
```

I should be able to deserialize all return type properly including the generic types.

### Actual Behaviour

Unfortunately, both methods: 
```kotlin
    @Get("/api/dummy/wrapped/list")
    fun wrappedList(): ApiResponse<List<Dummy>>


    @Get("/api/dummy/wrapped/nested")
    fun wrappedNested(): ApiResponse<ApiResponse<Dummy>>
```
fails.

The method `wrappedList()` fails with the following error `Error decoding HTTP response body: Error decoding stream for type [class com.micronaut.serde.ApiResponse]: Cannot deserialize raw list`.

The method `wrappedNested()` fails with the following error `java.lang.ClassCastException: class java.util.LinkedHashMap cannot be cast to class com.micronaut.serde.Dummy (java.util.LinkedHashMap is in module java.base of loader 'bootstrap'; com.micronaut.serde.Dummy is in unnamed module of loader 'app')`

### Steps To Reproduce

1. Clone [this reproducer](https://github.com/jntakpe/micronaut-generic-type-issues)
2. Run `gradle test`

### Environment Information

- MacBook Pro M1
- JDK 17 and 11
- Kotlin 1.8.20

### Example Application

https://github.com/jntakpe/micronaut-generic-type-issues

### Version

3.8.8

**Comment by allanrizza** [comment]
Try using
`@Get("/api/dummy/wrapped/list")
    fun wrappedList(): ApiResponse<List<Dummy>>` 

 instead of
 
 `@Get("/api/dummy/wrapped/list")
    fun wrappedList(): ApiResponse<List<Dummy>>`

**Comment by graemerocher** [comment]
This should work in Micronaut 4.x when it comes out

**Comment by yawkat** [comment]
should work with 4.x

**Comment by olevold** [comment]
Should work, but doesn't. I'm running Micronaut 4.2.2 and I'm struggling with exactly this.

**Comment by graemerocher** [comment]
hit the same issue recently, seems we are not realigning the types correctly

**Comment by graemerocher** [comment]
@dstepanov could you take a look?

**Comment by MikePryadko** [comment]
Hi everyone! Why is this issue closed? In Micronaut `4.10.10` problem still exists!

Outter class:
```java
@Serdeable
public record ApiResponse<T>(
        boolean success,
        T data,
        ApiError error
) {}
```
Inner class:
```java
@Serdeable
public record CursorPageResponse<T>(
        String cursor,
        List<T> data
) {}
```
Try to use:
```java
final Argument<CursorPageResponse> responseTypeInner = Argument.of(CursorPageResponse.class, MySimpleDTO.class);
final Argument<ApiResponse> responseType = Argument.of(ApiResponse.class, responseTypeInner);
final HttpResponse<R> response = httpClient.exchange(request, responseType, responseType);
final <ApiResponse<CursorPageResponse<MySimpleDTO>>> body = response.getBody(responseType).get();
log.debug("cursor: {}", response.data().cursor());
```
And get `ClassCastException`, because `body.data().getClass().getSimpleName() == "LinkedHashMap"` 😐 
```
Caused by: java.lang.ClassCastException: class java.util.LinkedHashMap cannot be cast to class myproject.model.CursorPageResponse
```
If bug is really fixed than what I'm doing wrong?

Thanks in advice!)
